### ISSUE TITLE
Allow CCS skip_unavailable to be overridden per search

### ISSUE THREAD

**Issue by droberts195** [elastic/elasticsearch#82407]
The CCS setting `skip_unavailable` silently ignores the inability to search an index on a remote cluster via CCS.

This is set per remote connection, not per search.  However, there are situations where it would be desirable for the setting to be different for different types of searches.

For an adhoc search using Discover where the user just wants to get a few random examples of messages of a particular type they may want the behaviour of `skip_unavailable=true`.  However, for automated periodic searches such as transforms and ML datafeed searches they may prefer that the search retries a bit later instead of silently tolerating unavailable data.  For these automated searches the problem of a subset of the data being ignored becomes clearly visible in the summarised results of the anomaly detection job or transformed data, and may then cause downstream problems as that summarised data is used in higher level decision-making.

Therefore it would be good if certain searches could opt to always be informed of unavailable remote data even if the global default was to skip it.

**Comment by dnhatn** [comment]
I've opened https://github.com/elastic/elasticsearch/pull/83542.

**Comment by javanna** [comment]
I see where this request comes from, but I am a bit hesitant on making this change, because it introduces yet another knob on the search request, complexity on understanding precedence between the two ways to set the flag, and most importantly I am not happy with the current behaviour of `skip_unavailable` (see #33915) and I would not want to extend it to the search request as-is. 

The search API exposes `ignore_unavailable` to affect how indices are resolved on the coordinating node and control whether unavailable indices should be ignored or not. This is propagated to the remote clusters too, although in some cases `skip_unavailable` is misused to achieve the same behaviour.

The search API also exposes `allow_partial_search_results` to control whether it should return partial results in case some shards are unavailable: this is for cases where the overall index is available (hence it is correctly resolved on the coordinating node) but some shards are unavailable when we go and execute the query on them. This is also propagated through CCS and the error returned is 503 - no_shard_available_action_exception.

I wonder if the motivation behind requesting this change is that when users set `skip_unavailable` to `true` for some of their remote clusters, they are no longer notified of any error, and flags like `allow_partial_search_results` and `ignore_unavailable` take no effect. A solution for that could be to address #33915, in short `skip_unavailable` should be limited to ignoring errors around availability of the whole remote cluster, and should not swallow other errors like unavailable shard when querying or unavailable index during index resolution.

Would this address the issue you encountered @droberts195 ?

**Comment by droberts195** [comment]
The scenario where this came up was where a user had set `skip_unavailable` for their remote clusters, then did a rolling upgrade of their entire ecosystem - rolling upgrades of all remote clusters followed by rolling upgrade of the central CCS cluster.  When they did this ML jobs in the central CCS cluster silently ignored unavailable data in the remote clusters while the remote clusters were doing their rolling upgrades.

> in short `skip_unavailable` should be limited to ignoring errors around availability of the whole remote cluster

I think that would probably help, _if_ doing a rolling restart of a remote cluster doesn't lead to the CCS cluster considering that remote cluster to be completely unavailable.  And I guess the answer to that might depend on how the remote connection has been configured.  For example in sniff mode I guess if there are multiple gateways selected then the connection should be able to tolerate a rolling restart.  But in proxy mode what happens if the single proxy is being restarted?

But then if there are many different ways of configuring the remote connections and some of them enable the CCS cluster to respond appropriately to rolling restarts of remote clusters and others don't then we should at least document what configurations are safe, and make sure the one we're using in our own Cloud service is safe.

The high level requirement is that a user who has set up CCS in a in a documented best practice way or in the way that our own Cloud service configures it doesn't silently lose search results when their remote clusters are doing rolling restarts providing they have set `allow_partial_search_results: false` and retry those searches if they fail due to partial results.

**Comment by javanna** [comment]
Thanks for the feedback! Agreed, it may be subtle to distinguish between shards being unavailable, entire nodes being down and the entire remote clusters being down. We will give this a try and see how far we get.

**Comment by javanna** [comment]
I've been circling around the idea of adapting the current `'skip_unavailable` behavior, and some feedback I got was that such a change can be perceived as a breaking change. 

Our docs are pretty clear around it in saying "Whether the remote cluster is skipped in case it is searched through a cross-cluster search request but none of its nodes are available." but the setting can also be seen from the perspective of a remote cluster being optional and not wanting to get any error if search results are not coming from it (the current behavior).

One thing we haven't mentioned above is that the search response includes the number of remote clusters that were skipped, hence skipping clusters does not happen silently. While it is true that setting `skip_unavailable` to `true` may lead to partial results although you set `allow_partial_search_results` to `false`, consumers are notified of this behavior and can act upon it.

To summarize, we could adapt the `skip_unavailable` behavior, but we should probably add a third option to make sure it is not seen as a breaking change.  The meaning of 'unavailable' here may still be subjective though. We could on the other hand update the docs to align it with the current behavior and better highlight that the number of skipped clusters is returned as part of the search response.

@droberts195 I understand that ML needs to retry the request in case there were partial results. Would it be feasible to do the same if there were skipped clusters reported in the response?

**Comment by dnhatn** [comment]
@javanna @droberts195 Thanks for the discussion. I think checking the skipped clusters is the most reasonable option. I will close my PR and fix the `unavailable` bug instead.

**Comment by javanna** [comment]
heya @dnhatn based on my understanding, if we rely on consumers to check skipped clusters and act accordingly, the current `skip_unavailable` behavior is ok. Changing it would break expectations of existing users, and I wonder then if we need a third option that tries to distinguish between unavailable shards and unavailable clusters. In short, if users want `allow_partial_results` and `ignore_unavailable` to take effect on the remote cluster, they need to set `skip_unavailable` to `false`. I would propose that we expand and clarify the docs on this. What do you think?

**Comment by dnhatn** [comment]
@javanna I think the [point-in-time](https://www.elastic.co/guide/en/elasticsearch/reference/current/point-in-time-api.html#point-in-time-api-example) API is also affected by this issue. We can fail requests with skipped clusters, but it might be a breaking change for some users with skip_unavailble. Users can't check if a PIT is complete or partial because we don't return the skipped clusters in PIT responses.

**Comment by javanna** [comment]
good point @dnhatn it sounds like we should unify the behaviour between search and pit then, if possible? Would it make sense to open a separate issue for this and expand on it?
